internal/run/run.go

```diff
@@ -591,6 +598,107 @@ func (w *Workflow) validateDocument(ctx context.Context, parentStep *WorkflowSte
 	return res, err
 }
 
+func (w *Workflow) snapshotSource(ctx context.Context, parentStep *WorkflowStep, namespaceName string, documentPath string) error {
+	hasSchemaRegistry, _ := auth.HasWorkspaceFeatureFlag(ctx, shared.FeatureFlagsSchemaRegistry)
+	if !hasSchemaRegistry {
+		return nil
+	}
+
+	pl := bundler.NewPipeline(&bundler.PipelineOptions{})
+	memfs := fsextras.NewMemFS()
+
+	registryStep := parentStep.NewSubstep("Tracking OpenAPI Changes")
+
+	registryStep.NewSubstep("Snapshotting OpenAPI Revision")
+
+	_, err := pl.Localize(ctx, memfs, bundler.LocalizeOptions{
+		DocumentPath: documentPath,
+	})
+	if err != nil {
+		return fmt.Errorf("error localizing openapi document: %w", err)
+	}
+
+	gitRepo, err := git.NewLocalRepository(w.projectDir)
+	if err != nil {
+		log.From(ctx).Debug("error sniffing git repository", zap.Error(err))
+	}
+
+	revision := ""
+	if gitRepo != nil {
+		revision, err = gitRepo.HeadHash()
+		if err != nil {
+			log.From(ctx).Debug("error sniffing head commit hash", zap.Error(err))
+		}
+	}
+
+	err = pl.BuildOCIImage(ctx, bundler.NewReadWriteFS(memfs, memfs), &bundler.OCIBuildOptions{
+		Tags:         []string{"latest"},
+		Reproducible: true,
+		Annotations: bundler.OCIAnnotations{
+			Revision: revision,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("error bundling openapi artifact: %w", err)
+	}
+
+	serverURL := auth.GetServerURL()
+	insecurePublish := false
+	if strings.HasPrefix(serverURL, "http://") {
+		insecurePublish = true
+	}
+
+	reg := strings.TrimPrefix(serverURL, "http://")
+	reg = strings.TrimPrefix(reg, "https://")
+
+	tags := []string{"latest"} // TODO: read from workflow.yaml
+
+	registryStep.NewSubstep("Storing OpenAPI Revision")
+	pushResult, err := pl.PushOCIImage(ctx, memfs, &bundler.OCIPushOptions{
+		Tags:     tags,
+		Registry: reg,
+		Access: ocicommon.NewRepositoryAccess(config.GetSpeakeasyAPIKey(), namespaceName, ocicommon.RepositoryAccessOptions{
+			Insecure: insecurePublish,
+		}),
+	})
+	if err != nil && !errors.Is(err, ocicommon.ErrAccessGated) {
+		return fmt.Errorf("error publishing openapi bundle to registry: %w", err)
+	}
+
+	registryStep.SucceedWorkflow()
+
+	var manifestDigest *string
+	var blobDigest *string
+	if pushResult.References != nil && len(pushResult.References) > 0 {
+		manifestDigestStr := pushResult.References[0].ManifestDescriptor.Digest.String()
+		manifestDigest = &manifestDigestStr
+		manifestLayers := pushResult.References[0].Manifest.Layers
+		for _, layer := range manifestLayers {
+			if layer.MediaType == ocicommon.MediaTypeOpenAPIBundleV0 {
+				blobDigestStr := layer.Digest.String()
+				blobDigest = &blobDigestStr
+				break
+			}
+		}
+	}
```

Comment:
So you've branch off of my work? I think we should wait for my stuff to merge first before merging this PR